### PR TITLE
Backports of hashtable fixes

### PR DESCRIPTION
**Pull request by ptomato**
https://phabricator.endlessm.com/T13987

### CODE REVIEW

**Comment by tchx84** [comment]
Did not find anything evidently wrong in the code, check tests passed, and other tested passed as well (e.g. our ekl tests). The main difference with these patches is that I no longer get segfaults when running emeus  ascii art examples.

Should someone else take a look?

File: gi/arg.cpp [hunk 1]
```diff
@@ -345,6 +345,149 @@ gjs_array_to_g_list(JSContext   *context,
...
+{
+    /* Don't use key/value destructor functions here, because we can't
+     * construct correct ones in general if the value type is complex.
+     * Rely on the type-aware g_argument_release functions. */
+
+    GITypeTag key_type = g_type_info_get_tag(key_param_info);
+
+    if (key_type == GI_TYPE_TAG_UTF8 || key_type == GI_TYPE_TAG_FILENAME)
+        return g_hash_table_new(g_str_hash, g_str_equal);
+    return g_hash_table_new(g_direct_hash, g_direct_equal);
```
Comment (ebassi) [inline gi/arg.cpp]: You should use `NULL, NULL`, since it's relatively faster (no PLT jump, not function call).

Comment (ebassi) [inline gi/arg.cpp]: Aren't you leaking the value you're allocating?

Comment (ptomato) [inline gi/arg.cpp]: It's freed in https://github.com/endlessm/gjs/pull/25/files#diff-5c8f10a5b663874949bc3dfa980a2c3eR3034

**Comment by ptomato** [comment]
According to the documentation `g_direct_hash` is still called when you pass `NULL`, so I left that as it was, but `g_direct_equal` is inlined, so I replaced that with `NULL`. Also posted to the upstream bug

**Comment by ebassi** [comment]
CR pass.

**Pull request merged**